app-ng/src/utilities.lisp

```diff
@@ -4,36 +4,76 @@
-string followed by optional args (as in FORMAT)."
-  (when (keywordp level-or-fmt-string)
-    ;; Sanity check that it's a valid log level at macroexpansion
-    ;; time.
-    (cl-syslog:get-priority level-or-fmt-string))
-  (if (keywordp level-or-fmt-string)
-      `(with-locked-log ()
-         (cl-syslog:format-log
-          *logger*
-          ',level-or-fmt-string
-          "~A~@?"
-          (session-info)
-          ,@fmt-string-or-args))
-      `(with-locked-log ()
-         (cl-syslog:format-log
-          *logger*
-          ':debug
-          "~A~@?"
-          (session-info)
-          ,level-or-fmt-string
-          ,@fmt-string-or-args))))
+(defun get-random-state (arg)
+  (etypecase arg
+    (null (qvm:seeded-random-state nil))
+    (unsigned-byte (qvm:seeded-random-state arg))))
+
+;; Stolen from HUNCHENTOOT::ISO-TIME
+(defun iso-time (&optional (time (get-universal-time)))
+  "Returns the universal time TIME as a string in full ISO format."
+  (multiple-value-bind (second minute hour date month year)
+      (decode-universal-time time)
+    (format nil "~4,'0d-~2,'0d-~2,'0d ~2,'0d:~2,'0d:~2,'0d"
+            year month date hour minute second)))
+
+(defun collect-memory-registers (qvm addresses)
+  "Return a HASH-TABLE of exactly the QVM memory registers requested in ADDRESSES.
+
+ADDRESSES is a HASH-TABLE where the keys are mem register names and the values are either T to indicate that we should return all indices for the corresponding register, or else a LIST of the desired indices for that register."
+  (let ((results (make-hash-table :test 'equal)))
+    (maphash (lambda (name indexes)
+               (cond
+                 ;; Give everything back.
+                 ((eq indexes t)
+                  (loop :with mv := (gethash name (qvm::classical-memories qvm))
+                        :for idx :below (qvm::memory-view-length mv)
+                        :collect (qvm:memory-ref qvm name idx) :into mem
+                        :finally (push mem (gethash name results))))
+                 ;; Give only some things back.
+                 ((alexandria:proper-list-p indexes)
+                  (loop :for idx :in indexes
+                        :collect (qvm:memory-ref qvm name idx) :into mem
+                        :finally (push mem (gethash name results))))
+                 (t
+                  (error "Invalid address parameter for memory named ~S." name))))
+             addresses)
+    results))
+
+(defun simulation-method->qvm-type (simulation-method &key pauli-noise-p)
+  "Return the QVM type for the given SIMULATION-METHOD and noise characteristics.
+
+PAULI-NOISE-P indicates the presence or absence of Pauli channel noise."
+  (ecase simulation-method
+    (pure-state (if pauli-noise-p
+                    'qvm:depolarizing-qvm
+                    'qvm:pure-state-qvm))
+    (full-density-matrix 'qvm:density-qvm)))
+
+(defun memory-required-for-qvm (simulation-method allocation-method num-qubits gate-noise
```

Comment:
How do we feel about this method of estimating QVM memory requirements? I'm ignoring: fixed per-object allocation overhead, differences between native/foreign allocation (if any), classical memory, and basically everything except quantum state.

I considered using sbcl-specific APIs to try and determine the "real" size of an instantiated QVM, but reconsidered because (a) not portable, (b) potentially fragile if it needs internal APIs, (3) possibly still need to deal with foreign / shared memory allocation separately anyway.

Reply:
You'll need another copy of the wavefunction also in the `pure-state-qvm` case *if* you're going to calculate expectations. Other than that, I think: (1) this seems to capture the dominant requirements and (2) it might be a good exercise (but possibly not for this PR) to query SBCL and have a clearer picture of the memory footprint.

Reply:
Good point about `perform-expectation`. IIUC, that copy of the wavefunction is ephemeral? I forgot to mention it here in this comment, but I do say in the doc string for `handle-qvm-memory-estimate` that the estimate is sort of a lower-bound and only includes memory allocated for the persistent state, i.e. ignoring anything allocated in the course of simulation. I was honestly thinking more along the lines of small lisp objects that get `cons`ed along the way, and hadn't considered larger allocations like temporary copies of the wavefunction.

Also agree regarding (2).

Reply:
I see. That's right, the copy of the wavefunction lives only within the scope of `%perform-expectation`.